references/depth/stereo/cascade_evaluation.py

```diff
@@ -0,0 +1,242 @@
+    
+    left_image_pyramid = [image_left]
+    right_image_pyramid = [image_right]
+    for idx in range(0, cascades - 1):
+        ds_factor = int(2 ** (idx + 1))
+        ds_shape = (image_left.shape[-2] // ds_factor, image_left.shape[-1] // ds_factor)
+        left_image_pyramid += F.interpolate(image_left, size=ds_shape, mode="bilinear", align_corners=True).unsqueeze(0)
+        right_image_pyramid += F.interpolate(image_right, size=ds_shape, mode="bilinear", align_corners=True).unsqueeze(0)
+        
+    flow_init = None
+    for left_image, right_image in zip(reversed(left_image_pyramid), reversed(right_image_pyramid)):
+        flow_pred = model(left_image, right_image, flow_init, iterations=iterations)
+        # flow pred is a list
+        flow_init = flow_pred[-1]
+        
+    return flow_init
+
+@torch.inference_mode()
+def _evaluate(
+    model, args, val_dataset, *, padder_mode, print_freq=10, writter=None, step=None, iterations=10, cascades=1, batch_size=None, header=None, save_images=False, save_path="",
+):
+    """Helper function to compute various metrics (epe, etc.) for a model on a given dataset.
+    We process as many samples as possible with ddp.
+    """
+    model.eval()
+    header = header or "Test:"
+    device = torch.device(args.device)
+    metric_logger = utils.MetricLogger(delimiter="  ")
+    
+    effective_batch_size = batch_size * args.world_size
+    can_do_distributed_test = effective_batch_size < len(val_dataset)
+    
+    if args.distributed and can_do_distributed_test:
+        sampler = torch.utils.data.distributed.DistributedSampler(val_dataset, shuffle=False, drop_last=True)
+    else:
+        sampler = torch.utils.data.SequentialSampler(val_dataset)
+    
+    val_loader = torch.utils.data.DataLoader(
+        val_dataset,
+        sampler=sampler,
+        batch_size=batch_size,
+        pin_memory=True,
+        num_workers=args.workers,
+    )
+
+    iterations = iterations or args.recurrent_updates
+
+    logger = utils.MetricLogger()
+    for meter_name in args.metrics:
+        logger.add_meter(meter_name, fmt="{global_avg:.4f}")
+    if "f1" not in args.metrics:
+        logger.add_meter("f1", fmt="{global_avg:.4f}")
+        
+    if can_do_distributed_test:
+        should_forward = True
+        module = model
+    else:
+        should_forward = args.rank == 0
+        # doing forward on rank 0 will trigger a synchronization
+        # which will not be fulfilled by non rank 0 processes
+        module = model.module
+
+    num_processed_samples = 0
+    with torch.cuda.amp.autocast(enabled=args.mixed_precision, dtype=torch.float16):
+        if should_forward:
+            batch_idx = 0
+            for blob in metric_logger.log_every(val_loader, print_freq, header):
+                image_left, image_right, disp_gt, valid_disp_mask = (x.to(device) for x in blob)
+                padder = utils.InputPadder(image_left.shape, mode=padder_mode)
```

Comment:
Question, can we put this `InputPadder` as part of transform (passed to dataset)?

Reply:
That would unfortunately not work since we have to be able to unpad the out as well after we have passed through the model. I believe another alternative would be to bind the padding utility to the model forward pass somehow.

Reply:
I think for your alternative, if we can add the padding on the model without specifying `padder_mode` then that should be fine. But if we need to specify `padder_mode`, then I think it is better to put it here as it is now

Reply:
One way to avoid specifying `padder_mode`  would be to have a default variable in the model constructors, and include a setter for them in case the user wants to change the padder mode after instantiating the model.

Reply:
Lets keep it as it is now, since if we add it to the models it also meant we need to add this padding behaviour to all stereo matching models